External Media: Add arrow keys navigation

WunderBart opened the pull request:
Adds arrow key navigation on media items in the external media browser modal. This brings it to have parity with the core Media Library image grid navigation.

![arrow-key-media-grid](https://user-images.githubusercontent.com/967608/84435242-1a049080-abf7-11ea-9e93-0ebd863a8729.gif)

#### Changes proposed in this Pull Request:
* Adds a `onKeyDown` prop to `MediaItem`
* Adds a `columns` ref and window resize listener to count and store the number of grid image columns. This is used to know where to send focus to when the up/down arrow keys are pressed.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to a post and add an Image or Gallery block
* **Arrow key presses _should not close the modal_**. Focus different elements (search input, search button, etc) and press the arrow keys. It should never close the modal. Arrow keys should operate normally inside inputs (like the search input).
* **Media Grid image items should be able to navigate with arrow key presses**
* **Resize the browser to less than or greater than 600px so the images restack.** The `columns` number should be updated and up/down arrow keys should bring you to the correct image.
* **All other functionality should still work** (Selecting images via click, space, enter; load more, etc)

obenland committed:
External Media: Render Media after authentication (#16044)

See https://github.com/Automattic/jetpack/pull/15717#pullrequestreview-418201700

obenland committed:
[not verified] Extend button margin to featured image (#16059)

obenland committed:
[not verified] Handle 401 responses (#16041)

See https://github.com/Automattic/jetpack/pull/15717#pullrequestreview-418201700

obenland committed:
External Media: Better reflect accepted media types (#16071)

obenland committed:
External Media: Account for multiple images (#16077)

See https://github.com/Automattic/jetpack/pull/16071#issuecomment-639367271

WunderBart committed:
External Media: Fix Modal focus loss on arrow keypress (#16055)

Co-authored-by: Marek Hrabe <marekhrabe@me.com>

WunderBart committed:
Switch media list to interactive container
...and remove the Load More button from its scope

WunderBart committed:
Create basic arrow keys navigation [WiP]

obenland committed:
External Media: Add API unit tests (#15988)

* [not verified] Add External media endpoint

* [not verified] Add unit tests

* [not verified] Add class coverage info

* Simplify schema

* Fix image copy test

* Use local image

* Remove duplicate file ending

Props @frontdevde.

obenland committed:
External Media: Improve Insert Flow (#16081)

* [not verified] # This is a combination of 5 commits.

External Media: Improve Insert Flow

See https://github.com/Automattic/jetpack/issues/15867

Remove unused import

Lock pointer events when copying
This prevents a UX issue where user can still interact with the modal content while the "Inserting..." request is pending.

Announce Inserting action along with alt text of selected images.

Update modal headers to reflect status

Props @jancavan.

* External Media: Fix mobile modal size (#16098)

* External Media: Improve Insert Flow

See https://github.com/Automattic/jetpack/issues/15867

* Remove unused import

* Lock pointer events when copying
This prevents a UX issue where user can still interact with the modal content while the "Inserting..." request is pending.

* Announce Inserting action along with alt text of selected images.

* Fix mobile mobile width to take up the full available modal space

* Change is-copying to modifier instead of element to follow BEM syntax

Co-authored-by: Konstantin Obenland <obenland@gmx.de>
Co-authored-by: Bart <bartlomiej.kalisz@gmail.com>

* External Media: Loading and disabled states for the copying state (#16103)

* disable pagination while loading

* fade out all images when copying

* disable event handlers and mark as disabled when copying

* disable pexels search when copying

* disabled google filters when copying

* add text to copying indicator

* disable google photos view selector when copying

* [not verified] remove debugging condition

* Pass functions not booleans to event emitters

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* Removed :not() copying css state since they are the same modal now

* Reset focus to the modal after clicking Insert. Otherwise focus is reset to the body and focus needs to stay trapped within the modal.

* Cosmetically rearranging code order

* External Media: Screen Reader accessible states for Insert flow (#16113)

* Add visually hidden description for the external media modal to give improved screen reader instructions and communicate the copying state

* [not verified] Remove title attribute from grid images

* [not verified] Move Inserting Image... copy to before the image so it is announced first.

* Separate-out description and element id

Helps with keeping describedby id in sync

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

Co-authored-by: Jerry Jones <jones.jeremydavid@gmail.com>
Co-authored-by: Bart <bartlomiej.kalisz@gmail.com>
Co-authored-by: Marek Hrabe <marekhrabe@me.com>

jeryj committed:
Changed event.stopPropogation to event.preventDefault for preventing the arrow key presses in the modal from bubbling to editor.

I'm not sure why this works. From my understanding, stopPropogation should stop the modal from bubbling up to the editor while allowing the keyDown listener on the media item. However, stopPropagation was blocking the media item keydown from being able to be called in the media grid.

jeryj committed:
Rework existing keydown listener to utilize prop on the media item rather than a listener on the wrapping element

jeryj committed:
Revert back to ul and li elements for the media items

jeryj committed:
Introduce columns state

jeryj committed:
Check how many images are in a row of the media grid and set this as the total columns for arrow navigation

WunderBart commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -99,36 +104,92 @@ function MediaBrowser( props ) {
+			const firstOffset = items[ 0 ].offsetTop;
+
+			// Check how many items have a matching offsetTop.
+			// This will give us the total number of items in a row.
```
Nice! 🙌

WunderBart commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -99,36 +104,92 @@ function MediaBrowser( props ) {
+			}
+		}
+
+		setColumns( perRow );
```
I think columns can be a mutable value as we need it only when pressing a key - no need for a state here.

getdave replied:
Avoid `setState` wherever possible as it will always cause a re-render. If you need the value to persist between component render lifecycles you can use a `ref` which won't cause a re-render when it's updated.

jeryj replied:
Still very much learning React, so thanks for the tips here!

WunderBart commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -99,36 +104,92 @@ function MediaBrowser( props ) {
+
+	useEffect( () => {
+		checkColumns();
+	}, [ media ] );
```
I think this check is good for setting the initial value. To prevent it from firing every time the media prop changes, we could do something like:

```js
useEffect( () => {
  if ( ! columns ) {
    checkColumns();
  }
}, [ media ] );
```

getdave replied:
In which case `columns` needs to be a dep of the effect right?

WunderBart replied:
Updated in https://github.com/Automattic/jetpack/pull/16118/commits/18e12bceb9c97fa17ae4f531cbbea638b6aae120. In that implementation columns value is set only once, when media are loaded.

WunderBart commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -99,36 +104,92 @@ function MediaBrowser( props ) {
+							setFocused( index );
+						} }
+						onKeyDown={ event => {
+							handleArrowKeysNavigation( event.keyCode, index );
```
I don't think we should attach the handler to every item. This can cause performance issues, especially when paired with the "Load More" button which will create a new handler for every item that will be added. Instead, we should create a single handler on the parent and delegate the event.

retrofox replied:
is it possible to use the useCallback hook?

getdave replied:
So my instinct is to say, yes delegation is better. Especially if we anticipate there are a tonne of media items. However, I'm not sure how much of an impact this will actually make.

Also there could be other knock-on issues caused by attempting this refactoring. For example, how will you get access to the `index` value if you attach the handler to the `<ul>` for example? 

Perhaps another option could be to use `useCallback` to memorize the result of the function for a given set of params. As per `useMemo` the function...

> ...will only recompute the memoized value when one of the dependencies has changed. 

...and therefore...

> This optimization helps to avoid expensive calculations on every render.

WunderBart replied:
> how will you get access to the index value if you attach the handler to the <ul> for example?

The index value is not necessary in this case, this works well based on the `focused` value only, as in https://github.com/Automattic/jetpack/pull/16118/commits/5fa86c8706f8e4719fb773999dbc8d03ae82221d#diff-f59fe14d5eeda67bd102566e399a16d4R104-R127

getdave replied:
Overall though I doubt the performance will be impacted much either way. Inline functions are very cheap.

WunderBart replied:
I guess if we agree that it **will** have a performance impact (no matter how big) we still should avoid it if possible. In this case, the solution is really just moving that handler to the parent so I think we still should do that.

jeryj replied:
While I still don't know React very well, at this point I've read plenty of Gutenberg and Jetpack React code. I haven't seen an instance where someone adds the keydown listener or click listener to the parent. It appears as though React pools events and that they are [synthetic events](https://reactjs.org/docs/events.html), so I don't think we need to worry about the performance impact.

Also, I added the event listener there because there was already a click listener and keydown listener there, so this code isn't actually adding any new event listener (as far as I know).

WunderBart replied:
Cleaned up & consolidated handlers for clarity in https://github.com/Automattic/jetpack/pull/16118/commits/827298d31cc534ab62bcf7fb3541cec55d98460a.

WunderBart commented on extensions/shared/external-media/media-browser/media-item.js:
```diff
@@ -18,16 +18,23 @@ function MediaItem( props ) {
+	const onKeyDown = useCallback(
+		event => {
+			// Catch space and enter key presses.
+			if ( ENTER === event.keyCode || SPACE === event.keyCode ) {
```
This whole handler could also be moved to parent component. No need to handle each item individually.

getdave replied:
See https://github.com/Automattic/jetpack/pull/16118#discussion_r438716986

getdave replied:
Is it me or will this function only be recomputed if `props.onKeyDown` changes? Would that mean that it will always memorze the first instance of its invocation? I think perhaps it only "works" because the `onKeyDown` is probably a different reference on each render. However if this was itself memorized then this could start to cause issues.

getdave replied:
Thinking about this perhaps it just returns the same function _reference_. However it doesn't memoize the _result_ of calling the function so maybe we're ok.

WunderBart commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -63,41 +65,79 @@ function MediaBrowser( props ) {
 	return (
 		<div className={ wrapper }>
-			<ul className={ classes }>
+			<div role="presentation" className={ classes } onKeyDown={ handleListKeyDown }>
```
I actually wanted to check-in with you on that 😄 
Assuming we will attach the handler to the `ul` element, the `eslint` will throw the [jsx-a11y/no-noninteractive-element-interactions](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-noninteractive-element-interactions.md) error. The guide suggests adding the `role="presentation"` attribute. What do you think would be the best course here?

jeryj replied:
`role="presentation"` is for saying "this element doesn't really matter, you can ignore it." Like, it's a decorative element that carries no further meaning. Definitely not the case here 😅 Disabling the error would be better in this instance.

WunderBart committed:
External Media: Restore focus on modal close (#16102)

WunderBart commented on extensions/shared/external-media/sources/with-media.js:
```diff
@@ -69,7 +69,7 @@ export default function withMedia() {
 				 */
 				if ( [ UP, DOWN, LEFT, RIGHT ].includes( event.keyCode ) ) {
-					event.stopPropagation();
+					event.preventDefault();
```
I'm afraid we can't use `preventDefault` here - it blocks the navigation inside i.e. the Pexels search input as it's one of the default `keydown` events.

getdave replied:
In which case we should probably add a comment to help future devs to be aware of this.

jeryj replied:
Ugh.  I knew it was too good to be true.

WunderBart replied:
Reverted in https://github.com/Automattic/jetpack/pull/16118/commits/0971fcf24dc015c6d143d36924856b6d3f06546c

WunderBart committed:
Clean up MediaBrowser & keep item handlers in one place

WunderBart committed:
Use ref for columns value

WunderBart committed:
Add missing prop

WunderBart committed:
Set columns value on window resize

obenland committed:
External Media: Prevent clicks from propagating out (#16124)

* [not verified] # This is a combination of 2 commits.

External Media: Prevent clicks from propagating out

Fixes #16070.

* Cap all propagation at MediaButton level

Prevents event propagation from clicks into the modal, including non-interactive elements.

WunderBart committed:
Fix typo

WunderBart committed:
Reword a comment

WunderBart committed:
Revert the stopPropagation for the arrow keys

WunderBart committed:
Submit selection on enter keydown rather than select item

jeryj commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -37,9 +58,12 @@ function MediaBrowser( props ) {
 		onCopy,
 	} = props;
 	const [ selected, setSelected ] = useState( [] );
+	const [ focused, setFocused ] = useState( -1 );
```
Should this be a useRef as well instead of in state?

WunderBart replied:
Change of focused value should cause a re-render, no? I don't think that a ref would work in this case.

jeryj replied:
I'm not sure. I'm not sure how a render is defined. It's not re-rendering anything in the HTML 🤷

WunderBart replied:
The `MediaItem` component visuals depend on the `focused` state, so the `focused` value change should trigger the `MediaItem` update. I hope that is true 😄

jeryj replied:
Tested this out with a ref instead. It only worked for the load more handler, likely because it was already triggering a rerender.

WunderBart committed:
Set columns value once when media are loaded

jeryj commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -55,14 +79,11 @@ function MediaBrowser( props ) {
 			}
 
 			setSelected( newSelected );
+			setFocused( index );
```
I think trying to keep track of what is currently focused in this way is going to end up being really difficult to get right in the end 😬 I don't think we need to keep track of it overall, just respond to the keydown listener from the media item.

WunderBart replied:
Updated in https://github.com/Automattic/jetpack/pull/16118/commits/9aff1e247def4b907bb29a63058ccda176bbb17d

jeryj commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -73,63 +94,133 @@ function MediaBrowser( props ) {
+
+	const checkColumnsDebounced = debounce( checkColumns, 400 );
+
+	useEffect( () => {
+		// Re-set columns on window resize:
+		window.addEventListener( 'resize', checkColumnsDebounced );
+		return () => {
+			window.removeEventListener( 'resize', checkColumnsDebounced );
+		};
+	}, [] ); // eslint-disable-line react-hooks/exhaustive-deps
```
So fresh and so clean 🙌

WunderBart committed:
Don't call setFocus on each select

WunderBart committed:
Don't render extra spinner
(Probably a leftover after a rebase)

WunderBart commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -168,24 +179,17 @@ function MediaBrowser( props ) {
-		if ( media.length ) {
-			setFocused( media.length );
+		if ( ENTER === event.keyCode ) {
+			handleSelectButtonClick();
```
This felt much more natural for me so I decided to go with it :)

jeryj committed:
Revert ENTER to submit on media items since that's not how the core media library handles it

jeryj committed:
Removed index from select() calls, as it was unused

jeryj committed:
Removed setFocused call on media item click, as focus will already be handled by the browser

jeryj commented on extensions/shared/external-media/media-browser/index.js:
```diff
@@ -37,8 +58,11 @@ function MediaBrowser( props ) {
 		onCopy,
 	} = props;
 	const [ selected, setSelected ] = useState( [] );
+	const [ focused, setFocused ] = useState( -1 );
```
I'm not sure what it should be named, but I'm worried calling this `focused` and `setFocused` indicates we're trying to keep track of what media item has focus. We don't actually care to keep track of the currently focused item. We only care about _sending focus to an item in some contexts_:

- Arrow key press
- Load More button click

The rest of the contexts (tab, click, etc) are already handled by the browser and we don't need to try to track or update those. I think we should rename this to something that is more in-line with what the state is doing: Sending Focus to a new item. Ideas:

- `sendFocusTo`: Accurate, but sounds like a method
- `nextFocus`: Accurate, but could have the same issue as above
- `navigateTo`: Similar to `sendFocusTo` but maybe sounds less like a method?

None of these feel great though. Any suggestions?

jeryj committed:
Renamed modalRef to contentRef because modalRef was getting stored as the contentElement and the parent of contentElement was called modalElement. Confusing.

jeryj commented on extensions/shared/external-media/sources/with-media.js:
```diff
@@ -68,7 +71,10 @@ export default function withMedia() {
-				if ( [ UP, DOWN, LEFT, RIGHT ].includes( event.keyCode ) ) {
+				if (
+					[ UP, DOWN, LEFT, RIGHT ].includes( event.keyCode ) &&
+					! this.contentElement.contains( event.target ) // Let event through to enable arrow keys navigation.
```
I found that this does not stop propagation on the pexels search input or search button. So when you press the arrow keys on those, propagation is not stopped and the modal closes. It doesn't close when using arrow keys on the media items because those call `stopPropagation` separately within `handleMediaItemKeyDown`. 

The search input and button do not have anything that calls `stopPropogation`, so the modal closes on arrow key presses. I think we should change this check to _only let arrow key presses avoid this `stopPropogation` if the event target is a media item_. I'm not sure a non-hacky way to do this though :/ It'd be nice to use refs, but I don't think that will be possible.

jeryj committed:
Only let arrow key presses avoid the initial stopPropogation call if the event.target is not one of our media grid items

jeryj commented on extensions/shared/external-media/sources/with-media.js:
```diff
@@ -68,7 +71,10 @@ export default function withMedia() {
-				if ( [ UP, DOWN, LEFT, RIGHT ].includes( event.keyCode ) ) {
+				if (
+					[ UP, DOWN, LEFT, RIGHT ].includes( event.keyCode ) &&
+					! event.target.classList.contains( 'jetpack-external-media-browser__media__item' ) // Only let arrow key navigation on media grid items through. All others need to be stopped.
```
I don't like that this uses a class check. The gist, [as explained earlier](https://github.com/Automattic/jetpack/pull/16118#discussion_r439028722) is that we can _only let arrow key presses on media grid items avoid this `stopPropogation` call_. 

The question is, **how do we guarantee that the `event.target` really is a media item**? 

- Is there a way to access its ref?
- Should we use the HTML instead, like a `li` or `role="checkbox"` check? This seems more likely to get missed though.

jeryj commented on extensions/shared/external-media/sources/with-media.js:
```diff
@@ -37,8 +37,10 @@ export default function withMedia() {
+			contentRef = el => {
 				if ( el ) {
+					// Store modal content.
+					this.contentElement = el;
```
Depending on what happens about [the stopPropogation issue](https://github.com/Automattic/jetpack/pull/16118/files#r439038767), we may not need this `contentElement` at all.

jeryj commented:
Closing in favor of https://github.com/Automattic/jetpack/pull/16794, as this one is quite out of date and has a lot of extraneous work from other branches in it.
